components/script/dom/xmlhttprequest.rs

```diff
@@ -1347,29 +1347,18 @@ impl XMLHttpRequest {
         blob
     }
 
-    // https://xhr.spec.whatwg.org/#arraybuffer-response
-    #[allow(unsafe_code)]
-    fn arraybuffer_response(&self, cx: JSContext) -> Option<NonNull<JSObject>> {
+    /// <https://xhr.spec.whatwg.org/#arraybuffer-response>
+    fn arraybuffer_response(&self, cx: JSContext) -> Option<ArrayBuffer> {
         // Step 1
-        let created = self.response_arraybuffer.get();
-        if let Some(nonnull) = NonNull::new(created) {
-            return Some(nonnull);
-        }
-
-        // Step 2
-        let bytes = self.response.borrow();
-        rooted!(in(*cx) let mut array_buffer = ptr::null_mut::<JSObject>());
-        unsafe {
-            ArrayBuffer::create(*cx, CreateWith::Slice(&bytes), array_buffer.handle_mut())
-                .ok()
-                .and_then(|()| {
-                    self.response_arraybuffer.set(array_buffer.get());
-                    Some(NonNull::new_unchecked(array_buffer.get()))
-                })
+        if !self.response_arraybuffer.is_initialized() {
+            // Step 2
+            let bytes = self.response.borrow();
+            self.response_arraybuffer.set_data(cx, &bytes).ok()?;
         }
+        self.response_arraybuffer.get_buffer().ok()
```

Comment:
However, now I'm not sure the step comments make much sense. Looking into the spec it says:

```
5. If this’s response type is "arraybuffer", then set this’s response object to a new ArrayBuffer object representing this’s received bytes. If this throws an exception, then set this’s response object to failure and return null. 
```

I'm not sure what the two substeps were exactly referring to. Does it make sense to leave them like this, should I move them somewhere else or remove them?

Reply:
Yeah...this is an issue. There was a tendency to only include the step numbers in the code until now, but the spec renumbers the operations all the time. I would just try to implement what the specification says now and don't be afraid to renumber things (but this time include a description of the step with the number).

Reply:
That makes a lot of sense. Would something like this last commit work?

Reply:
Yes, that looks great. Thank you!

Reply:
Oh, wait. I'm sorry, I spoke to soon. The step numbers and text should correspond to what is in the specification at https://xhr.spec.whatwg.org/#arraybuffer-response.

Reply:
I'm guessing what happened here is that an older version of the specification had this separated into different sections but the latest version has it in a single one -- which is why the step numbers don't really make sense any longer.

Reply:
Yes, I figured it was something like that. Should I simply remove the step numbers and leave the comments as they are then? Or change the top comment to `Step 5:` and remove the other step?

Reply:
That makes sense to me. Thanks!

Reply:
Thank you! It should be ready now